File: noaa-requests/matlab/NOAA_Water_Level_Downloader/WL_Downloader_Functions/date_search.m
function [indx, dEnd, dBeg, invalid_flag] = date_search(dStruct,dBeg,dEnd,indx)
% Flag
invalid_flag = [];
% Get Date Ranges
RangeStart = dStruct.startDate(indx);% Start
RangeEnd = dStruct.endDate(indx);% End
% Cut Date Strings
RangeStart = cellfun(@(x) datenum(x(1:end-4),'yyyy-mm-dd HH:MM:SS'),RangeStart,'UniformOutput',false);
RangeEnd = cellfun(@(x) datenum(x(1:end-4),'yyyy-mm-dd HH:MM:SS'),RangeEnd,'UniformOutput',false);
% Check If Query Dates Are Within Product Date Ranges (A<x<B, where x->query dates)
%     uBound_chk_dummy = cell2mat(cellfun(@(x) cell2mat(RangeStart)<=x & x<=cell2mat(RangeEnd),{dBeg},'UniformOutput',false));
uBound_chk_dummy = cell2mat(cellfun(@(x) x>=cell2mat(RangeStart) & x<cell2mat(RangeEnd),{dBeg},'UniformOutput',false));
lBound_chk_dummy = cell2mat(cellfun(@(x) x>cell2mat(RangeStart) & x<=cell2mat(RangeEnd),{dEnd},'UniformOutput',false));
% Logical Vector (A<x<B, where x->query dates)
dummy_indx = uBound_chk_dummy+lBound_chk_dummy;
if all(dummy_indx==0) % Requested Date Range Is Outside Data Availability
    % % Find Nearest Entry To Requested End Date
    % [~, indx2] = min(abs(cell2mat(RangeEnd)-dEnd));
    % % Overwrite Requested Date Range With Best Option
    % dEnd = RangeEnd{indx2}; % Start
    % dBeg = RangeStart{indx2}; % End
    % % Get Row index
    % indx = indx(indx2);
    invalid_flag = 1;
elseif any(dummy_indx == 2) % Requested Date Range Is Withing Data Prodcuts Availability
    indx = indx(dummy_indx == 2);
elseif any(dummy_indx == 1) % Requested Date Range Needs Adjustment On One Of the Extremes
    % Determine Anchor
    if all(uBound_chk_dummy == 0) % Requested Start Date Is Within Range
        % Find Nearest Entry To Requested End Date
        [~, indx2] = min(abs(cell2mat(RangeStart)-dBeg));
        % Find Corresponding End Date
        dBeg = RangeStart{indx2};
        % Get Row index
        indx = indx(indx2);
    elseif all(lBound_chk_dummy == 0)
        % Find Nearest Entry To Requested End Date
        [~, indx2] = min(abs(cell2mat(RangeEnd)-dEnd));
        % Find Corresponding End Date
        dEnd = RangeEnd{indx2};
        % Get Row index
        indx = indx(indx2);
    end
else
    invalid_flag = 1;
end
end

File: conversion/noaa-requests/noaa-py/noaapy/date/date_search.py
def date_search(d_struct, d_beg, d_end, indx):
    invalid_flag = False

    range_start = d_struct['start_date'][indx]
    range_end = d_struct['end_date'][indx]

    range_start = [datetime.datetime.strptime(item[:19], "%Y-%m-%d %H:%M:%S") for item in range_start]
    range_end = [datetime.datetime.strptime(item[:19], "%Y-%m-%d %H:%M:%S") for item in range_end]

    u_bound_chk_dummy = [d_beg >= start and d_beg <= end for start, end in zip(range_start, range_end)]
    l_bound_chk_dummy = [d_end >= start and d_end <= end for start, end in zip(range_start, range_end)]

    dummy_indx = [u + l for u, l in zip(u_bound_chk_dummy, l_bound_chk_dummy)]

    if all(item == 0 for item in dummy_indx):
        invalid_flag = True
    elif any(item == 2 for item in dummy_indx):
        indx = [i for i, value in enumerate(dummy_indx) if value == 2]
    elif any(item == 1 for item in dummy_indx):
        if all(u == 0 for u in u_bound_chk_dummy):
            nearest_start = min(range_start, key=lambda x: abs(x - d_beg))
            d_beg = nearest_start
        elif all(l == 0 for l in l_bound_chk_dummy):
            nearest_end = min(range_end, key=lambda x: abs(x - d_end))
            d_end = nearest_end

    return indx, d_end, d_beg, invalid_flag

